[SPARK-7127] [MLLIB] Adding broadcast of model before prediction for ensembles

BryanCutler opened the pull request:
Broadcast of ensemble models in transformImpl before call to predict

BryanCutler committed:
[SPARK-7127] Adding broadcast of model before prediction in RandomForestClassifier

AmplabJenkins commented:
Can one of the admins verify this patch?

BryanCutler commented on mllib/src/main/scala/org/apache/spark/ml/Predictor.scala:
```diff
@@ -184,9 +185,45 @@ abstract class PredictionModel[FeaturesType, M <: PredictionModel[FeaturesType,
+  ]
+  extends PredictionModel[FeaturesType, M] {
+
+  protected def transformImpl(dataset: DataFrame, bcastModel: Broadcast[M]): DataFrame = {
```
I'm not sure if using the Broadcast variable as a parameter is a good idea

BryanCutler commented on mllib/src/main/scala/org/apache/spark/ml/classification/RandomForestClassifier.scala:
```diff
@@ -134,13 +136,28 @@ final class RandomForestClassificationModel private[ml] (
-    // TODO: Override transform() to broadcast model.  SPARK-7127
     // TODO: When we add a generic Bagging class, handle transform there: SPARK-7128
+    // Predict without using a broadcasted mode
+    predictImpl(features, () => this)
```
Do we still want to support prediction without broadcasting?

jkbradley replied:
Yes, we will want to.  If it's a small model, then it can be faster to not broadcast it.

jkbradley commented:
Thanks for the PR!  I don't think we should introduce a new abstraction.  The abstractions can be useful, but it seems like a lot of overhead for the small task of broadcasting models.  It will also complicate things since we can't have multiple inheritance from 2 abstract classes.  Can you please remove it for now?

BryanCutler committed:
[SPARK-7127] Removed abstract class for broadcasting model, instead passing a prediction function as param to transform

BryanCutler commented:
I agree @jkbradley and removed the new class.  I changed this around a little in the last commit and I think it's a cleaner.  Now `transformImpl`  takes the predict function as a parameter, then the overridden `transform` in a concrete ensemble classifier/regressor will broadcast the model and bind the broadcast var to the `predictImpl` where it is accessed.  Let me know if this design seems ok and I'll implement for the other ensembles.  Thanks!

BryanCutler commented on mllib/src/main/scala/org/apache/spark/ml/classification/RandomForestClassifier.scala:
```diff
@@ -134,13 +135,23 @@ final class RandomForestClassificationModel private[ml] (
   override def treeWeights: Array[Double] = _treeWeights
 
+  override def transform(dataset: DataFrame): DataFrame = {
+    val bcastModel = dataset.sqlContext.sparkContext.broadcast(this)
```
You mentioned that we might want to selectively broadcast the model, only if it's large enough.  Do you think that is something we can do here automatically, or would it need to be a configuration setting?

BryanCutler commented:
Hi @jkbradley, just wondering if you could take a look at the changes from my last commit and see if they are ok.  Thanks!

BryanCutler committed:
[SPARK-7127] Applied broadcasting to remaining ensemble models

BryanCutler committed:
[SPARK-7127] Used modelAccessor parameter in predictImpl to access broadcasted model

BryanCutler committed:
[SPARK-7127] Removed accidental newline

jkbradley commented:
@BryanCutler Sorry for the delay!  I like the general idea, but I think it could be simpler.  What if:
* Predictor.transform still handled everything, except the actual prediction.  For that, it would call transformImpl(dataset).  I'll note what I mean inline.
* Predictor.transformImpl would by default use predict(), as before.
* Subclasses like RandomForestClassifier could override transformImpl to broadcast the model and then use that broadcast variable in a map (which would use predict()).

That should allow you to do the same thing, but you can have subclasses not override transform() and can eliminate predictImpl.  (Also, currently, the subclasses skip schema validation in transform, which is a problem.)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/Predictor.scala:
```diff
@@ -174,9 +174,15 @@ abstract class PredictionModel[FeaturesType, M <: PredictionModel[FeaturesType,
+      predictFunc: (FeaturesType) => Double): DataFrame = {
     transformSchema(dataset.schema, logging = true)
     if ($(predictionCol).nonEmpty) {
-      dataset.withColumn($(predictionCol), callUDF(predict _, DoubleType, col($(featuresCol))))
```
This could call:
```
transformImpl(dataset)
```

BryanCutler commented:
Hi @jkbradley , thanks for checking this out!  I'm not sure I understand a couple things from your suggestion.

If a subclass implements `transformImpl(dataset: DataFrame)`, then broadcasts and proceeds with a`dataset.map(...)` then the result is now an RDD which would have to be made back into a DataFrame to return.  This seems like an inefficient step, which is why I tried to just stick with DataFrames.

Also, the model parameters need to be accessed from inside `predict(features: Double)` of a subclass, like RandomForestClassificationModel, so the only way to do this is to change the signature of `predict` to have the broadcast var as a parameter, or make the broadcast var a member of RandomForestClassifier.  Both of those seemed like bad ideas, which is why I added `predictImpl` that could share the same code for both broadcasted and non-broadcasted models.

Sorry, maybe I am missing something, could you elaborate more on how you were thinking of using the broadcast variable in a map?

jkbradley commented:
Sorry, I should not have said "map."  I agree we should not use map since that will create an RDD.  By map, I really meant UDF.

Here's the sketch of what I meant for transform/predict:
* Predictor.transform: check schema, and call transformImpl
* Predictor.transformImpl: call predict in a UDF, as it does now
* RF/GBT*.predict: keep as is
* RF/GBT*.transformImpl (override): broadcast model, and call UDF.  Inside the UDF, get model from the broadcast variable, and call predict on it.

Does that make sense?

AmplabJenkins commented:
Can one of the admins verify this patch?

BryanCutler committed:
[SPARK-7127] Simplified calls by overriding transformImpl and using broadcasted model in callUDF to make prediction

BryanCutler committed:
Merge branch 'master' into bcast-ensemble-models-7127

Conflicts:
	mllib/src/main/scala/org/apache/spark/ml/Predictor.scala

BryanCutler committed:
[SPARK-7127] Replaced deprecated callUDF with udf

BryanCutler commented:
Yeah, that makes sense.  It wasn't clear to me that using the broadcast model in the UDF would allow the member vars in `predict` to later use that broadcast instance.  Hopefully this last iteration is more along the lines of what you were thinking.  I also merged with the latest master and replaced `callUDF` with `udf`.  Thanks for the help @jkbradley !

jkbradley commented:
Jenkins test this please

jkbradley commented:
@BryanCutler Nice, the updates look good, and it's a bit simpler now.  LGTM pending tests.

SparkQA commented:
[Test build #1077 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1077/consoleFull) for   PR 6300 at commit [`86e73de`](https://github.com/apache/spark/commit/86e73ded458c8cc8790c13a98c4a0f3f7efb7dbc).

SparkQA commented:
[Test build #1077 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1077/console) for   PR 6300 at commit [`86e73de`](https://github.com/apache/spark/commit/86e73ded458c8cc8790c13a98c4a0f3f7efb7dbc).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented:
Merging with master.
Thank you for the contribution!

BryanCutler commented:
Cool, thanks!

Tamilselvan-Veeramani commented:
can any one help me on, how to use transformImpl method for predictProbability method ? I see it's not implemented in transformImpl of RandomForestClassificationModel. hence my streaming job broad casting the RF model for every mini batch. Help me with way to implement. thanks
